---
title: "FNA transcriptomics"
author: "Ed Carr, Linterman lab"
date: "8 Dec 2020"
output: 
  pdf_document: 
editor_options: 
  chunk_output_type: console
---

```{r echo=F}
Sys.setenv(PATH=paste(Sys.getenv("PATH"),
                      "/bi/home/carre/texlive/2017/bin/x86_64-linux/",sep=":"))

```

```{r setup, include=FALSE}

## set a few options for pdf formatting.
knitr::opts_chunk$set(echo = F, tidy = T, warning = F, message = F, fig.width = 7, fig.height = 9.5, tidy.opts=list(width.cutoff=60))
```

```{r libraries}

# As this Rmd has grown, and the slow (many Gb) chunks are not evaluated each run,
# we should explicitly load libraries.
library(Seurat)
library(ggplot2)
library(cowplot)
library(patchwork)

```

```{r read_into_Seurat, echo=TRUE, eval=FALSE}

library(Seurat)
########
# Wrapper function
# Read in each sample, annotate with day
SRRs2Seurat <- function(x) {
  data <- Read10X(data.dir = as.character(x$path))
  obj <- CreateSeuratObject(data, 
                            project = 
                              "FNA")
                            
  obj <- AddMetaData(obj, metadata = c(as.character(x$day), 
                                       x$lib,
                                       paste(
                                         as.character(x$day), 
                                    x$lib, sep = "_")
                                    ),
                     col.name = c("day", "lib","day_lib"))
  
  # % mito reads
  obj[["percent.mt"]] <- PercentageFeatureSet(obj, pattern = "^MT-")
  
  # simple QC: >200 features/cell, less than 7k features/cell and <12.5% mitochondrial reads.
  obj <- subset(obj, subset = nFeature_RNA > 200 & nFeature_RNA < 7000 & percent.mt < 12.5)
}
#########

# Build a dataframe with sample annotation:

SRRtable <- cbind("path" = c("cellranger/TurnerSRR11233645/outs/filtered_feature_bc_matrix/",
                             "cellranger/TurnerSRR11233646/outs/filtered_feature_bc_matrix/",
                             "cellranger/TurnerSRR11233647/outs/filtered_feature_bc_matrix/",
                             "cellranger/TurnerSRR11233648/outs/filtered_feature_bc_matrix/",
                             "cellranger/TurnerSRR11233649/outs/filtered_feature_bc_matrix/",
                             "cellranger/TurnerSRR11233650/outs/filtered_feature_bc_matrix/",
                             "cellranger/TurnerSRR11233651/outs/filtered_feature_bc_matrix/",
                             "cellranger/TurnerSRR11233653/outs/filtered_feature_bc_matrix/",
                             "cellranger/TurnerSRR11233654/outs/filtered_feature_bc_matrix/",
                             "cellranger/TurnerSRR11233655/outs/filtered_feature_bc_matrix/",
                             "cellranger/TurnerSRR11233656/outs/filtered_feature_bc_matrix/"),
                  "day" = c("d60",
                            "d60",
                            "d28",
                            "d28",
                            "d12",
                            "d12",
                            "d12",
                            "d5",
                            "d5",
                            "d0",
                            "d0"),
                  # SRA lists FNA_2 or _3 (presumably each core is a #)
                  # And lib 1 or 2 (presumambly 10x channel replicates)
                  # Only some of these combinations exist
                  # Presumably some cores were too low cell number to process.
                  # Presumably some 10x channels clearly failed in prep prior to sequencing.
                  # these are simplified into (1, 2) here.
                  "lib" = c(1,
                            2,
                            2,
                            1,
                            3,
                            2,
                            1,
                            2,
                            1,
                            2,
                            1))


SRRtable <- data.frame(SRRtable)
SRRtable



# Make a list
fna.list <- list()

# Loop for each entry in SRR table to do the import to Seurat.
for(i in 1:nrow(SRRtable)){

  fna.list[[i]] <- SRRs2Seurat(SRRtable[i,])
  
}


names(fna.list) <- paste(SRRtable$day, SRRtable$lib, sep = "_")
```


```{r logNorm_integrationWithAnchors, echo=TRUE, eval=FALSE}
# First time running, this chunk needs eval=TRUE
# This is a slow chunk, so run once normalised data and written to disk.
# hence eval=FALSE

for (i in 1:length(fna.list)) {
    fna.list[[i]] <- NormalizeData(fna.list[[i]], verbose = FALSE)
    fna.list[[i]] <- FindVariableFeatures(fna.list[[i]], selection.method = "vst", 
        nfeatures = 2000, verbose = FALSE)
}


# Set reference as a day 12 sample (will have some GC-B, and all other cell types)
# (the Seurat vignette keeps one dataset back to 'project' anchors onto)
reference.list <- which (names(fna.list) == "d12_1")

# Find anchors:
fna.anchors <- FindIntegrationAnchors(object.list = fna.list, normalization.method = "LogNormalize", reference = reference.list, dims = 1:30)

fna.integrated <- IntegrateData(anchorset = fna.anchors, dims = 1:30)

system("mkdir seurat_objects")

save(fna.integrated, file = "seurat_objects/FNA_transcriptomes_AnchorIntegrated.RData")
```

```{r echo=TRUE, eval=FALSE}


load("seurat_objects/FNA_transcriptomes_AnchorIntegrated.RData")

# switch to integrated assay. The variable features of this assay are automatically
# set during IntegrateData
DefaultAssay(fna.integrated) <- "integrated"

# Run the standard workflow for visualization and clustering
fna.integrated <- ScaleData(fna.integrated, verbose = FALSE)
fna.integrated <- RunPCA(fna.integrated, npcs = 20, verbose = FALSE)

fna.integrated <- FindNeighbors(fna.integrated, dims = 1:20)
fna.integrated <- FindClusters(fna.integrated, resolution = 0.05)

fna.integrated <- RunUMAP(fna.integrated, reduction = "pca", dims = 1:20)
p1 <- DimPlot(fna.integrated, reduction = "umap", group.by = "day")
p2 <- DimPlot(fna.integrated, reduction = "umap", group.by = "seurat_clusters")
#p1 + p2

```

```{r echo=TRUE, eval=FALSE}

# identify B cells
# Should use the 'RNA' assay for visualisation like this.

DefaultAssay(fna.integrated) <- "RNA"
# Normalize RNA data for visualization purposes
fna.integrated <- NormalizeData(fna.integrated, verbose = FALSE)

## Turner et al. use these markers for PBMC/FNA cluster assignment:
markers.to.plot <- c("MS4A1", "CD19", "CD79A","CD3D", "CD3E", "CD3G", "IL7R", "CD4", "CD8A","GZMB", "GNLY", "NCAM1", "CD14", "LYZ", "CST3", "MS4A7", "IL3RA", "CLEC4C", "PPBP")


p3 <- DotPlot(fna.integrated, features = c(markers.to.plot)) + ggpubr::rotate_x_text()

# Clusters 1, 4 and 6 are B cells
fna.B <- subset(fna.integrated, idents = c("1", "4", "6"))

# Turner et al check for housekeeping genes:
#DotPlot(fna.integrated, features = c("ACTB", "GAPDH", "B2M", "HSP90AB1", "GUSB", "PPIH", "PGK1", "TBP", "TFRC", "SDHA", "LDHA"))

# Save B cells
DefaultAssay(fna.B) <- "integrated" # set explicitly, so we re-cluster on integrated anchors.
save(fna.B, file = "seurat_objects/1_FNA_transcriptomes_AnchorIntegrated_just_B_cells.RData")

```

```{r}

load("seurat_objects/1_FNA_transcriptomes_AnchorIntegrated_just_B_cells.RData")

DefaultAssay(fna.B) <- "integrated" # set explicitly, so we re-cluster on integrated anchors.

# Run the standard workflow for visualization and clustering
fna.B <- ScaleData(fna.B, verbose = FALSE)
fna.B <- RunPCA(fna.B, npcs = 20, verbose = FALSE)

fna.B <- FindNeighbors(fna.B, dims = 1:20)
fna.B <- FindClusters(fna.B, resolution = 0.05) # Turner used 0.2, for this integration approach 0.05 gives biologically relevant clusters.

fna.B <- RunUMAP(fna.B, reduction = "pca", dims = 1:20)

b1 <- DimPlot(fna.B, reduction = "umap", split.by = "day")
b2 <- DimPlot(fna.B, reduction = "umap", group.by = "seurat_clusters")
b1 + b2

# Turner definitions
gcb.markers <- c("BCL6", "RGS13", "MEF2B", "STMN1", "ELL3", "SERPINA9")


DefaultAssay(fna.B) <- "RNA"
b3 <- DotPlot(fna.B, features = c( "TCL1A", "IL4R", "CCR7", "IGHM", "IGHD",
                                   gcb.markers, 
                                     "XBP1", "IRF4", "SEC11C", "FKBP11", "JCHAIN", "PRDM1", 
                                   "TNFRSF13B", "CD24", "CD27", "TBX21", "FCRL5", "ITGAX", "NKG7", "ZEB2", "CR2")) + ggpubr::rotate_x_text()
b2
b1 + b2 + b3
#b2 + b3
#FeaturePlot(fna.B, features = "CXCR4")
cowplot::plot_grid(b2,b3, rel_widths = c(1,2))

library(tidyverse)
as_tibble(fna.B[[]]) %>% 
  mutate("day" = factor(day, levels = c("d0","d5","d12","d28","d60"))) %>% 
  ggplot(aes(day,  fill = seurat_clusters)) + geom_bar(position = "fill") + ylab("Proportion") + ggpubr::theme_pubr()
```

```{r}
# Save GC B cell barcodes:

## A couple of different approaches here.
## There are, expectedly, very few day 0 and day 5 GC B cells.
## Options:
## 1. use all GC B cell barcodes
## 2. use day 12 GC B cells alone (the d0 GC B cells will NOT be QIV specific).
## 3. use day 28 GC B cells alone

# Strategy 1:
## Use GC B cells from day 12 as the 'ground truth'. Track these BCRs into blood AND GCBs in FNA at later time points.
GCbarcodes <- subset(fna.B, subset = day == "d12") %>% WhichCells(., idents = "3")
save(GCbarcodes, file = "seurat_objects/1_FNA_GC_B_d12_barcodes.RData")
# 
```

# SessionInfo

```{r sessionInfo}

sessionInfo()

```