---
title: ""
author: ""
date: ""
output: 
  pdf_document: 
    latex_engine: lualatex
papersize: a4
header-includes:
    - \pagenumbering{gobble}
editor_options: 
  chunk_output_type: console
---

```{r echo=F, eval=F}
#Sys.setenv(PATH=paste(Sys.getenv("PATH"),
 #                     "/bi/home/carre/texlive/2017/bin/x86_64-linux/",sep=":"))

```

```{r setup, include=FALSE}

## set a few options for pdf formatting.
knitr::opts_chunk$set(echo = F, tidy = T, warning = F, message = F, fig.width = 7.5, fig.height = 11.5, tidy.opts=list(width.cutoff=60))
```

```{r load_data}

library(MultiAssayExperiment)
library(ggplot2)

library(magrittr)
library(dplyr)
library(ggpubr)

library(pals)
library(cowplot)

library(SingleCellExperiment)
library(scater)
library(scran)

library(ggcyto)
library(rstatix)

load(file = "../cohort_2016_17/ALFNA16_MultiAssayExperiment.RData")
load(file = "../cohort_2016_17/data/SCE_QC_pass_finalised.RData")
load(file = "../cohort_2016_17/Example_of_sc_sort_gating_from_665X_d42.RData")


```


```{r study_design_panel}


study.layout <-
  cbind("Assay"=c(
    "Haemagglutinin\ninhibition assay",
    "Haemagglutinin\ninhibition assay",
    "Haemagglutinin\ninhibition assay",
    "Haemagglutinin\nELISPOT",
    "Index sorting of\nhaemagglutinin-\nspecific B cells",
    "Index sorting of\nhaemagglutinin-\nspecific B cells"),
    
    "day"=as.numeric(c(0,7,14,7,0,14)))

study.layout <- as.data.frame(study.layout)
study.layout$day <- as.numeric(as.character(study.layout$day))
study.layout$Assay <- factor(study.layout$Assay, 
                             levels =c(
                                       "Haemagglutinin\ninhibition assay",
                                       "Haemagglutinin\nELISPOT",
                                       "Index sorting of\nhaemagglutinin-\nspecific B cells", ""))

study <- ggplot(study.layout, aes(x=day, y=Assay, col = Assay,fill=Assay)) + geom_point(shape=25, size=3) + 
  scale_x_continuous(breaks=c(0,7,14), labels=c(0,7,42))+
  theme_bw() + theme(legend.position = "none", 
                     panel.grid = element_blank(), 
                     panel.border = element_blank(), 
                     text = element_text(size=12),
                     axis.title.y = element_blank()) +
  annotate(geom= "segment",x = 0,yend = 4.2,xend = 0, y= 4.5, size = 1, arrow = arrow(length = unit(6, "points"))) +
  annotate(geom="text", x=1, y=4.35, label="Trivalent flu vaccine", hjust="left") +
  labs(subtitle = "Healthy volunteers\n22-36 years old (n=10)\n67-86 years old (n=10)")

```

```{r hai}
hai.data <- wideFormat(ALFNA16[ , , 
                                # select assays with names that match:
                                # 'HAI'
                                # which returns all data from these two assays at all available days.
                                grepl(names(assays(ALFNA16)), pattern = "HAI") ],
                      # Return ages as well:
                       colDataCols = "Age.Group")


hai.data <- as.data.frame(hai.data)

hai.data <- hai.data %>% dplyr::filter(primary %in% sce$PID)

```


```{r run_SingleR}

library(SingleR)
mon.se <- MonacoImmuneData()

counts <- logNormCounts(sce)


# Change rownames from ENSG ID to symbols:
library(org.Hs.eg.db)
rownames(counts) <- mapIds(
  org.Hs.eg.db,
  keys=rownames(counts),
  keytype="ENSEMBL", column="SYMBOL", multiVals = "first")

# Remove any NA gene symbols:
counts <- counts[ !isNA(rownames(counts)), ]

# Predict - note we are using the 'fine' labels:
pred <- SingleR(test = counts, ref = list(MON=mon.se), 
                labels = list(MON=mon.se$label.fine))

pred$shortlabels <- pred$labels
pred$shortlabels %<>% gsub(., pattern = "Exhausted B cells", replacement = "IgD-CD27- Bmem")
pred$shortlabels %<>% gsub(., pattern = "Naive B cells", replacement = "Naive")
pred$shortlabels %<>% gsub(., pattern = "Non-switched memory B cells", replacement = "Non-switched Bmem")
pred$shortlabels %<>% gsub(., pattern = "Switched memory B cells", replacement = "Switched Bmem")
pred$shortlabels %<>% gsub(., pattern = "Plasmablasts", replacement = "Plasmablast")
pred$shortlabels %<>% gsub(., pattern = "Plasmacytoid dendritic cells", replacement = "pDC")

pred$shortlabels <- factor(pred$shortlabels,
                           levels = c("Naive",
                                      "Non-switched Bmem",
                                      "Switched Bmem",
                                      "IgD-CD27- Bmem",
                                      "Plasmablast",
                                      "pDC")) # a levels argument forces order for plots.


sce$labels <- pred$shortlabels


```

```{r}
## flow % HA panel.

library(tidyverse)
library(ggpubr) # for themes for plots.


load(file = "../cohort_2016_17/Concatenated_index_master_fcs_gated_in_FlowJo_2_table_output.RData")

merged.data <- as_tibble(merged.data)
merged.data$y1 <- merged.data$`Lymphocytes/Single Cells/Single Cells/DUMP-/CD19+SA-/hA_APC+hA_PE+ | Freq. of Parent (%)`

y1 <- "y1"
ylab1 <- "\nHA+ B cells \n[% of live B cells]"

data.for.ggpaired1 <- reshape2::dcast(
  merged.data, PID+age ~ day, value.var = y1)

data.for.ggpaired1$expansion <- data.for.ggpaired1$d42/data.for.ggpaired1$d0

# data.for.ggpaired1 %>%
#   group_by(age) %>%
#   dplyr::summarise(median = median(expansion))

```



```{r day7_PBs}

load("../cohort_2016_17/ALFNA16_MultiAssayExperiment.RData")

elispot <- wideFormat(ALFNA16[ , , 
                                # select assays with names that match:
                                # 'HAI'
                                # which returns all data from these two assays at all available days.
                                grepl(names(assays(ALFNA16)), pattern = "ELISPOT") ],
                      # Return ages as well:
                       colDataCols = "Age.Group")




elispot$Age <- factor(c("22-36yo", "67-86yo")[elispot$Age.Group])

elispot <- as.data.frame(elispot)
# cowplot::plot_grid(
# ggboxplot(elispot, x = "Age", y = "ELISPOT_d7_HA..per.1e6.PBMC",
#           ylab = "HA ELISPOT day 7 per 10^6 PBMC") + stat_compare_means(method ="wilcox", paired=F, comparisons = list(c(1,2))),
# ggboxplot(elispot, x = "Age", y = "ELISPOT_d7_Fluvax..per.1e6.PBMC",
#           ylab = "Fluvax ELISPOT day 7 per 10^6 PBMC") + stat_compare_means(method ="wilcox", paired=F, comparisons = list(c(1,2))),
# ncol = 2)


elispot <- elispot %>% dplyr::filter(primary %in% sce$PID)

```



```{r flow validation}
flowCytoDat.HA_PB <- openxlsx::read.xlsx(xlsxFile = "../cohort_2016_17/data/FlowValidation/Fig1_FlowCytoValidation.xlsx",
                                   sheet = "HA_plasmablasts_of_Bcells")


HAPBpanel <- flowCytoDat.HA_PB %>%
  mutate(ID = paste(Group, ID)) %>%
  pivot_longer(contains("day"),
               names_to = "day",
               values_to = "%") %>%
 mutate(day = str_remove(day, "ay")) %>%
  mutate(day = factor(day, levels = c("d0", "d7","d42"))) %>%
  mutate(Group = case_when(Group=="Young"~ "22-36yo",
                           Group=="Old"~"67-86yo", 
                           is.na(Group)==TRUE ~"NA")) %>%
  mutate(Group  = factor(Group, levels = c("22-36yo", "67-86yo"))) %>%
  mutate(`%`= as.numeric(`%`)) %>%
  ggplot(aes(x=Group, y=`%`,group=Group)) + 
  geom_boxplot(aes(group=NULL, col = Group), outlier.shape = NA) + 
  geom_point(aes(group=Group, fill=Group), shape=21,
             size=2,position=position_jitter(width = 0.3)) +
  scale_fill_manual(values = c("white", "darkgrey")) +
  scale_color_manual(values = c("black", "black")) + 
  facet_grid(.~day) + 
  ggpubr::stat_compare_means(comparisons = list(c(1,2)),label.y = 0.03, method="wilcox") + 
  labs(y="HA-specific\nplasmablasts [% of B cells]") + 
  ggpubr::theme_pubr(legend = "none") +
  theme(axis.title.x = element_blank(),strip.background = element_rect(fill = "NA")) + rotate_x_text()


#----------#

flowCytoDat.PB <- openxlsx::read.xlsx(xlsxFile = "../cohort_2016_17/data/FlowValidation/Fig1_FlowCytoValidation.xlsx",
                                   sheet = "Plasmablasts_of_Bcells")


  PBpanel <- flowCytoDat.PB %>%
  mutate(ID = paste(Group, ID)) %>%
  pivot_longer(contains("day"),
               names_to = "day",
               values_to = "%") %>%
  mutate(day = str_remove(day, "ay")) %>%
  mutate(day = factor(day, levels = c("d0", "d7","d42"))) %>%
    mutate(Group = case_when(Group=="Young"~ "22-36yo",
                           Group=="Old"~"67-86yo", 
                           is.na(Group)==TRUE ~"NA")) %>%
  mutate(Group  = factor(Group, levels = c("22-36yo", "67-86yo"))) %>%
  mutate(`%`= as.numeric(`%`)) %>%
  ggplot(aes(x=Group, y=`%`,group=Group)) + 
  geom_boxplot(aes(group=NULL, col = Group), outlier.shape = NA) + 
  geom_point(aes(group=Group, fill=Group), shape=21,
             size=2,position=position_jitter(width = 0.3)) +
  scale_fill_manual(values = c("white", "darkgrey")) +
  scale_color_manual(values = c("black", "black")) + 
  facet_grid(.~day) + 
  ggpubr::stat_compare_means(comparisons = list(c(1,2)),label.y=4, method="wilcox") + 
  labs(y="Plasmablasts\n[% of B cells]") + 
  ggpubr::theme_pubr(legend = "none") +
  theme(axis.title.x = element_blank(),strip.background = element_rect(fill = "NA")) + rotate_x_text()

  
ggsave(cowplot::plot_grid(HAPBpanel, PBpanel, ncol=2), file = "Aurora_PBs.svg", width=6, height=4,dpi=300)
```


```{r plot_figure}

symnum.args <- list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
                    symbols = c("****", "***", "**", "*", "ns")) # for the P values 



hai.p.values <- hai.data %>%
  dplyr::select(primary, Age.Group, 
                -contains("HAI_d"), contains("log2")) %>%
  mutate(Age.Group=str_replace_all(Age.Group, c("young"="18-22yo", "old"="67-86yo"))) %>%
               mutate("Age" = factor(paste0(Age.Group))) %>%
  dplyr::select(-Age.Group) %>%
  set_colnames(
    gsub(colnames(.), 
         pattern = "HAI_d|_log2.titre", replacement = "")) %>%
  reshape2::melt( id_vars = c(primary, Age)) %>%
  set_colnames(gsub(colnames(.), pattern = "variable", replacement = "day")) %>%
  set_colnames(gsub(colnames(.), pattern = "value", replacement = "titre")) %>%
  arrange(., Age, primary) %>%
  group_by(Age) %>%
  wilcox_test(titre ~ day, paired = T)

hai.p.values <- hai.p.values %>% 
  add_xy_position(x = "day") %>% 
  add_significance(p.col = "p", 
                   cutpoints = symnum.args$cutpoints,
                   symbols = symnum.args$symbols)





plot_grid(
  # ROW 1:
  plot_grid(
    # study design
    study,
    
    # HAI titres, faceted by age.
    # HAI titres, faceted by age.
ggplot(data = 
               hai.data %>%
               dplyr::select(primary, Age.Group, 
                             contains("HAI_d"), -contains("log2")) %>%
         mutate(Age.Group=str_replace_all(Age.Group, c("young"="18-22yo", "old"="67-86yo"))) %>%
               mutate("Age" = factor(paste0(Age.Group))) %>%
               dplyr::select(-Age.Group) %>%
               set_colnames(
                 gsub(colnames(.), 
                      pattern = "HAI_d|_titre", replacement = "")) %>%
               reshape2::melt( id_vars = c(primary, Age)),
       aes(x = variable, y=value,fill = Age, group=primary)) +
         geom_violin(aes(fill=NULL, group=NULL), col="grey", trim=F) +
    geom_line(position = position_dodge(width=0.4), col = "lightgrey", lwd=0.5) +
  geom_point(shape =21, position = position_dodge(width=0.4)) +
  scale_fill_manual(values = c("white", "darkgrey")) +
  labs(x ="day", y="A/Cal09 HAI titer") +
  stat_pvalue_manual(hai.p.values, label = "p.signif", step.increase = c(0.1)) +
  scale_y_continuous(trans = "log2", limits = c(1, 10000)) +

  facet_grid(.~Age) +
  theme_pubr(legend = "none") +
  theme(strip.background = element_rect(fill = "NA")),
    
    ### Day 7 ELISPOTS ###
    ggboxplot(elispot, x = "Age", y = "ELISPOT_d7_HA..per.1e6.PBMC",
          ylab = "HA ELISPOT d7 per 10<sup>6</sup> PBMC") + stat_compare_means(method ="wilcox", paired=F, comparisons = list(c(1,2))) + 
      ylim(0,600) +
      xlab(NULL)  + rotate_x_text() + 
  theme(axis.title.y = ggtext::element_markdown()),
ggboxplot(elispot, x = "Age", y = "ELISPOT_d7_Fluvax..per.1e6.PBMC",
            ylab = "Fluvax d7 ELISPOT per 10<sup>6</sup> PBMC") + stat_compare_means(method ="wilcox", paired=F, comparisons = list(c(1,2))) + 
  ylim(0,3000) +
  xlab(NULL) + rotate_x_text() + 
  theme(axis.title.y = ggtext::element_markdown()),
    labels = LETTERS[1:4],
    rel_widths = c(3,2,1.3,1.3),
    ncol = 4),
  
  # ROW 2:
  plot_grid(
    
    # gtable of a ggcyto plot
    fig,

    ## This is the HA % of live B cells.
    ##
    ggpaired(data.for.ggpaired1, cond1 = "d0", cond2 = "d42", facet.by = "age",
           ylab = ylab1, line.color = "grey",
           xlab = "day") +
    stat_compare_means(method = "wilcox",
                       label.y = 0.35,
                       paired = TRUE,
                       comparisons = list(c(1,2))) +
      theme_pubr() +
    theme(strip.background = element_rect(fill = NA)),
    
    
  # UMAP both day 0 and day 42
  # Louvain clustering
  plotUMAP(sce, colour_by="clusters") + 
    theme_pubr(legend = "top") + 
    guides(fill = guide_legend(title = "Louvain\ncluster",
                               title.position = "left",ncol=2)) + 
    theme(legend.key.size = unit(0, 'lines'),
          legend.margin = margin(0,0,0,0, 'lines'),
          aspect.ratio = 1) + rotate_x_text(),
  
  
  labels = LETTERS[5:7],
  rel_widths = c(1.5,2,2),
  ncol = 3),
  
  # ROW 3:
  plot_grid(
    # LEFT HAND PANEL:
  # UMAP both day 0 and day 42
  # Colours from SingleR labels.
  plotUMAP(sce[ , sce$labels != "pDC"], colour_by="labels") +
    scale_fill_brewer(palette = "Spectral") +
    theme_pubr() +
    labs(fill =  "cell\nassignment") +
    theme(legend.position = "bottom", legend.direction = "vertical",
          axis.ticks = element_blank(),
          axis.text = element_blank(),
          legend.key.size = unit(0, 'lines'),
          legend.margin = margin(0,0,0,0, 'lines'),
          aspect.ratio = 1),
    
    # RIGHT HAND 4 UMAPs:
    plot_grid(
  # UMAP both day 0 and day 42
  # Colours from index flow IgD
  plotUMAP(sce, colour_by="IgD.BUV737") +
    viridis::scale_fill_viridis(option="inferno") + 
    labs(fill = "Surface\nIgD") +
    theme_pubr() + 
    theme(legend.position = "right",
          axis.ticks = element_blank(),
          axis.text = element_blank(),
          legend.key.width = unit(6,"points"),
          legend.key.height = unit(12,"points"),
          legend.margin = margin(0,0,0,0, 'lines'),
          aspect.ratio = 1),
  
   # UMAP both day 0 and day 42
  # Colours from index flow CD27 
    plotUMAP(sce, colour_by="CD27.BV711") + 
    viridis::scale_fill_viridis(option="inferno") +
     labs(fill = "Surface\nCD27") +
     theme_pubr() +
     theme(legend.position = "right",
           axis.ticks = element_blank(),
           axis.text = element_blank(),
           legend.key.width = unit(6,"points"),
           legend.key.height = unit(12,"points"),
           legend.margin = margin(0,0,0,0, 'lines'),
           aspect.ratio = 1),
  
  # UMAP both day 0 and day 42
  # Colours from index flow CD21
  plotUMAP(sce, colour_by="CD21.PE.cy7") +
    viridis::scale_fill_viridis(option="inferno") +
    labs(fill = "Surface\nCD21") +
    theme_pubr() +
    theme(legend.position = "right",
          axis.ticks = element_blank(),
          axis.text = element_blank(),
          legend.key.width = unit(6,"points"),
          legend.key.height = unit(12,"points"),
          legend.margin = margin(0,0,0,0, 'lines'),
          aspect.ratio = 1),
  
  # UMAP both day 0 and day 42
  # Colours from index flow CD28
  plotUMAP(sce, colour_by="CD38.BV421") + 
    viridis::scale_fill_viridis(option="inferno") +
    labs(fill = "Surface\nCD38") +
    theme_pubr() +
    theme(legend.position = "right",
          axis.ticks = element_blank(),
          axis.text = element_blank(),
          legend.key.width = unit(6,"points"),
          legend.key.height = unit(12,"points"),
          legend.margin = margin(0,0,0,0, 'lines'),
          aspect.ratio = 1),
  
  labels = LETTERS[9:12],
  ncol = 2),
  
  labels = c(LETTERS[8], NULL),
  rel_widths = c(1,2),
  ncol = 2),
  ncol = 1, rel_heights = c(1.5,1.25,1.5))


```

__Figure 1 – Single cell sequencing of haemagluttinin specific B cells to study the aged vaccine response__


\newpage

# SessionInfo

```{r sessionInfo}

sessionInfo()

```